pkg/destroy/libvirt/connection/connection.go

```diff
@@ -0,0 +1,111 @@
+//go:build libvirt
+// +build libvirt
+
+package connection
+
+import (
+	"crypto/tls"
+	"fmt"
+	"net"
+	"net/url"
+	"strings"
+	"time"
+
+	libvirt "github.com/digitalocean/go-libvirt"
+	"github.com/sirupsen/logrus"
+)
+
+// NewConnection creates a new connection to a libvirt daemon.
+func NewConnection(libvirtURI string, log logrus.FieldLogger) (conn *libvirt.Libvirt, disconnect func(), err error) {
+	disconnect = func() {
+		if conn != nil {
+			if err := conn.Disconnect(); err != nil && log != nil {
+				log.Warn("failed to disconnect from libvirt", err)
+			}
+		}
+	}
+
+	uri, err := url.Parse(libvirtURI)
+	if err != nil {
+		return
+	}
+
+	c, err := dial(uri)
+	if err != nil {
+		err = fmt.Errorf("could not open connection to libvirt: %w", err)
+		return
+	}
+
+	conn = libvirt.New(c)
+	err = connect(uri, conn, log)
+	return
+}
+
+func dial(uri *url.URL) (net.Conn, error) {
+	transport := "unix"
+	if scheme := strings.Split(uri.Scheme, "+"); len(scheme) > 1 {
+		transport = scheme[1]
+	} else if uri.Host != "" {
+		transport = "tls"
+	}
+
+	switch transport {
+	case "unix":
+		sock := "/var/run/libvirt/libvirt-sock"
+		if s := uri.Query().Get("socket"); s != "" {
+			sock = s
+		}
+		return net.DialTimeout(transport, sock, 2*time.Second)
```

Comment:
The dialers are added in a newer version of go-libvirt, and I was trying to keep to the same version that is used in the Terraform provider to minimise issues. At the time my understanding of the plugin system was that if we used the same dependency in both the core installer module and one of the Terraform plugins, we had to use the same version. But my latest understanding is that at least since @jhixson74's changes to run Terraform as a separate binary, that's no longer a constraint.

I did consider reusing the connection code from the terraform plugin, but the team has gone to a _lot_ of effort to remove all of those as dependencies, and I don't think we should add them back. The ideal would be if go-libvirt provided the functionality of this whole connection package, but as of now it doesn't.

Reply:
> The dialers are added in a newer version of go-libvirt, and I was trying to keep to the same version that is used in the Terraform provider to minimise issues. 

Ah fair enough.

> I did consider reusing the connection code from the terraform plugin, but the team has gone to a lot of effort to remove all of those as dependencies, and I don't think we should add them back. The ideal would be if go-libvirt provided the functionality of this whole connection package, but as of now it doesn't.

I also did not have this background information :) Have you considered copy&pasting https://github.com/dmacvicar/terraform-provider-libvirt/tree/main/libvirt/uri in the installer? Or does this come with too many dependencies?
That's not a blocker, but this looks like the kind of code which is going to have subtle differences in corner cases depending on the implementation you use, which is why I'm checking if we can avoid reimplementing it :)

Reply:
The unix and tcp connection types that we actually use are trivial enough that it doesn't matter.
I didn't originally plan to implement the tls one, but it seemed like it might be useful to be safe. Probably I should submit something like this upstream to go-libvirt.

Reply:
Yep, it's the tls code I had in mind with "subtle differences and corner cases". I agree that the other ones are trivial enough.

Reply:
Upstream PR: digitalocean/go-libvirt#209